### DISCUSSION TITLE
FES2022 released!

### DISCUSSION THREAD

**Discussion by robbibt** [pyTMD/pyTMD#309] [General]
Very exciting update today: FES2022 has finally been released! 🎉

Product page: https://www.aviso.altimetry.fr/en/data/products/auxiliary-products/global-tide-fes.html

User guide: https://www.aviso.altimetry.fr/fileadmin/documents/data/tools/hdbk_FES2022.pdf

Seems like some pretty nice improvements over coastal areas particularly compared to FES2014!

![Screenshot_20240712-213534.png](https://github.com/user-attachments/assets/455a958b-26bd-464f-8450-bb402a558c18)

![Screenshot_20240712-213549.png](https://github.com/user-attachments/assets/8da1b887-e140-4d53-8e8c-f45bb8e4d91c)

**Comment by robbibt** [comment]
Woah, just saw that there is already support for this - amazing! 🤯
https://github.com/tsutterley/pyTMD/releases/tag/2.1.2

**Comment by tsutterley** [comment]
Yep! Super excited to support FES2022!

**Comment by robbibt** [comment]
Hey @tsutterley, trying this out for the first time, I seem to be having some issues calculating tides from FES2022. This works fine for FES2014:
```
from pyTMD.compute import tide_elevations
import pandas as pd

tide_elevations(
    x=[122.21],
    y=[-18.20],
    delta_time=pd.date_range(start="2022-09-01", end="2022-09-30", freq="1h"),
    DIRECTORY="/gdata1/data/tide_models/",
    MODEL="FES2014",
    EPSG=4326,
    TIME="datetime",
)
```
![image](https://github.com/user-attachments/assets/dab2f580-e0fd-47a9-84d8-1bff8e1feff3)

However if I try the same with FES2022, I get `TypeError: ufunc 'bitwise_and' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''`:
```
tide_elevations(
    x=[122.21],
    y=[-18.20],
    delta_time=pd.date_range(start="2022-09-01", end="2022-09-30", freq="1h"),
    DIRECTORY="/gdata1/data/tide_models/",
    MODEL="FES2022",
    EPSG=4326,
    TIME="datetime",
)
```
![image](https://github.com/user-attachments/assets/4b2309d4-c4f2-4406-89f6-a8273d0e35b0)


I downloaded the FES2022 files with [aviso-fes-tides-py](https://pytmd.readthedocs.io/en/latest/api_reference/aviso_fes_tides.html#aviso-fes-tides-py), and my directory looks like this:
![image](https://github.com/user-attachments/assets/89842d75-f86d-44ed-a113-2dde89b4929f)
![image](https://github.com/user-attachments/assets/1c245d89-14a6-4538-ba44-ccbd2bd95cd8)

I'm on the latest version:
![image](https://github.com/user-attachments/assets/e1d76dae-f352-4da5-9b3d-4268e874f013)

**Reply by tsutterley** [reply]
Tracked this down (even though it is fixed in pyTMD :face_with_spiral_eyes:). The FES2022 files have an improperly set `valid_max` attribute. Should be 360 but is set to 0. I could fix it in the download script, but I'm undecided if it is worth the effort.

**Comment by tsutterley** [comment]
Hey @robbibt, super weird. For some reason the FES2022 netCDF files have the longitude arrays masked. I have a fix in #311.

**Reply by robbibt** [reply]
Fantastic, seems to be working so far! I'll do a bit of a validation and share results here. Thanks again for the quick fix!
![image](https://github.com/user-attachments/assets/d529ecc8-5082-40f3-a2c5-126b0c40cc28)

**Reply by tsutterley** [reply]
Awesome! Looking forward to seeing the results!

**Reply by robbibt** [reply]
Looks pretty encouraging so far! Results against ~100 GESLA tide guages across Australia, comparing FES2022 to seven other models for 2017-2019. We have a new winner!

![image](https://github.com/user-attachments/assets/f22f06cf-f29e-4212-a727-a6b7c9c1eef6)

**Reply by robbibt** [reply]
Hey @tsutterley, this is all working really nicely so far. I was wondering if it would be possible to get a new version released at some point with the masking fix incorporated? I'd love to be able to upgrade PyTMD in our workflows without having to check out via Github. 🙂

**Reply by tsutterley** [reply]
For sure I'll do it tomorrow 👍👍

**Reply by robbibt** [reply]
One last request (should have mentioned this yesterday!) - if we could get a conda update too that would be fantastic - the last one there is 2.1.1. 🙂 
https://anaconda.org/conda-forge/pytmd

**Reply by tsutterley** [reply]
for sure I'm just waiting for the PR from conda-forge to merge. I can do it manually tomorrow if I don't get an update for the recipe.

**Reply by tsutterley** [reply]
sometimes you just have to do things manually :laughing: conda is bumped to 2.1.3

**Reply by robbibt** [reply]
Awesome, thanks!

**Comment by robbibt** [comment]
Just two related questions after exploring FES2022 a little more: 

1. Compared to previous releases, it looks like FES2022 has done a funky thing with their "extrapolated" layers, where they have injected "0" values for inland lakes to prevent the models from running over those environments. Annoyingly for our applications, this also includes coastal wetland ecosystems that we most want to model! 😂 I just wanted to verify that PyTMD is able to handle these correctly given they're coded as "0" values and not standard nodata?

![image](https://github.com/user-attachments/assets/d320ffde-74cd-46ff-89d0-dfeacfb5f1e6)

2. Currently I've been testing differences between the standard and "extrapolated" versions of FES2022 by manually renaming the "ocean_tide_extrapolated" folder to "ocean_tide". Is there an easier way via the PyTMD API? (I know there's the "extrapolate" param, but I believe that's for when points fall completely out of the modelling domain, not whether or not to use the pre-generated extrapolated data)

**Reply by robbibt** [reply]
Just some further analysis: I suspect that pyTMD isn't currently accounting for those masked-out zero values. I've compared results from the standard FES2022 data against the "extrapolated" data for some of our tide gauges that sit on top of those masked pixels (e.g. GESLA Wyndham site code 63090; 128.101, -15.4533):
![image](https://github.com/user-attachments/assets/ddc3ccf8-e414-4f31-8428-1fd7e5dcc6b2)
![image](https://github.com/user-attachments/assets/f30867f4-511b-4578-a97d-2e8565855901)

And we seem to be getting back highly dampened tide amplitudes in the "extrapolated" data compared to the standard data:
![image](https://github.com/user-attachments/assets/aad45b94-2ffc-47d2-a139-509e81a2453c)

Even if there's a pyTMD work-around, I might also raise this issue with AVISO - hardcoding those "lakes" in the constituent files seems to defeat one of the main purposes of the "extrapolated" data (being able to generate a best-estimate of tides outside of the modelling domain). Especially annoying when many of those "lakes" aren't lakes at all, but tide-influenced coastal ecosystems!

**Reply by tsutterley** [reply]
Hey @robbibt,
Thanks for being my FES2022 beta tester :sweat_smile: :facepalm: sorry for the bumpy start!

I agree with you, I'd rather not zero out these inland and coastal water bodies with the extrapolated FES data. So in #314 I'm setting the mask for these values so that that they can be rightfully interpolated (when using the extrapolation routines within pyTMD).

Good question. I currently support models of a known type (FES/GOT/OTIS etc) that aren't in `io.model` by using definition files. I didn't really want to "support" FES2022-extrapolated out of the box I wanted to do more work to investigate the model. But for this case, I'm including the definition files for the FES2022-extrapolated model in #314 (in both [tab-delimited](https://github.com/tsutterley/pyTMD/blob/2a16ea7b51488885ada7ce1e68a94debb32af613/test/model_FES2022b_extrapolated.def) and [JSON](https://github.com/tsutterley/pyTMD/blob/2a16ea7b51488885ada7ce1e68a94debb32af613/test/model_FES2022b_extrapolated.json) formats). These use the "glob" functionality to read all the model files supplied in the tide model directory. The only change you should have to do on your end is use `--definition-file model_FES2022b_extrapolated.def` or `DEFINITION_FILE='model_FES2022b_extrapolated.def'` instead of `--tide FES2022` or `MODEL='FES2022'`. Still include the directory variable as before. 

The other recent update (from a PR merged yesterday) is to enable "on-the-fly" cropping of the model files. This should greatly improve the computational times for some of these higher resolution models. I think I have covered the prime meridian/dateline cases. If you want to test those out, you can set `--crop` or `CROP=True`. Hope this helps :+1:

**Reply by robbibt** [reply]
No worries, super appreciative to have FES2022 support so early - I think we're all learning about its quirks as we go! Thanks for the masking fix, I'll test the new code out on Monday. 

The other two changes sound fantastic too: we already do an initial manual crop on our end (running our analysis on copies of the files cropped to Australia), but the on-the-fly cropping should make it far faster (from what I've gathered, the slowest part is actually the interpolation vs. loading the constituent files, and the interpolation runs far faster on smaller inputs).

JSON inputs for definition files sounds great! Would this potentially make it possible to define definitions "in code", e.g. pass in something like a Python dictionary containing all required params, rather than having to load the definition from file? That would be really helpful for reproducibility - at the moment we store some custom def files (e.g. for FES2012) next to our model data, but that only works for our usage, not for external users who might want to run our code on their systems.

**Reply by robbibt** [reply]
OK, looks like the latest version does a great job of resolving issues over those masked lakes: we're no longer seeing artificially dampened tides after the fix:

![image](https://github.com/user-attachments/assets/525d0a88-8722-4f75-8ecf-623af5be45bd)

Compared against GESLA tide gauges, the "extrapolated" model now produces results that are very close to the standard model:

Before:
![image](https://github.com/user-attachments/assets/2e891c49-e955-4749-8c0d-835844ecd0ab)

After:
![image](https://github.com/user-attachments/assets/c91e1da7-10e7-4407-9460-a1874855befb)

**Comment by tsutterley** [comment]
Yes! Defining the JSON files as in-memory `io.BytesIO` objects is definitely possible and I think should already work if you `json.dump` your dictionary into one. That was at least the goal. Also yes, moving towards just being able to define everything as `kwargs`. Just moving slowly :joy: 

Ok perfect. Glad the dampening issue got fixed. THANK you for pointing it out! Was an interesting decision by the FES team to zero out the coastal water bodies.

**Comment by tsutterley** [comment]
based on your tests, should I put out a new release (2.1.4) so you can get things done?

**Reply by robbibt** [reply]
No rush, but that would be fantastic if you get a chance. It's all working nicely for me so far!

**Reply by tsutterley** [reply]
on it :+1:

**Comment by fowlerovski** [comment]
I'm trying to work with FES2022b as well. 
I can't seem to get a realistic load_tide calculation for the tidal loading of coastal regions (surface displacement). 

I've tried a number of different approaches, but either I end up with values of over 1 meter (not feasible for Florida), or sub millimeter, which also doesn't seem reasonable.

Any pro tips or tricks for computing OTL for topographic calculations? 

Thanks kindly!

**Reply by tsutterley** [reply]
Hi @fowlerovski, 
I'll look into it! Can you outline what you've tried? e.g. are you using the [`pyTMD.compute.tide_elevations`](https://pytmd.readthedocs.io/en/latest/api_reference/compute.html#pyTMD.compute.tide_elevations) command using the `FES2022_load` model?

**Reply by fowlerovski** [reply]
Hi @tsutterley, 
To be honest, I'm out of my depth with tidal calculations. 

I'm using pyTMD.io.model(tide_dir).elevation('FES2022_load')

I'm a bit embarrassed and will go read the docs some more so I have a clue what I'm asking here... 
I'll be back when I have better clue of what I'm doing. 

BTW, another question (with more validity): are there any standard check points for validating results?
I fiddle bits and get different results, with no feeling for ground truth... 

Is there a small set of coordinates/timestamps with precomputed values that can be used to validate against?

Thanks so much, great work!

**Reply by tsutterley** [reply]
hi @fowlerovski,
I'm still downloading the load tide files locally so that I can check what is happening :face_with_spiral_eyes:

You can compare ocean tide values against the [FES2022 test file](https://github.com/CNES/aviso-fes/blob/main/examples/fes_slev_2022_v2.9.7.txt):
```
import pyTMD
import pandas
df = pandas.read_csv('fes_slev_2022_v2.9.7.txt', delimiter=r'\s+')
h = 100.0*pyTMD.compute.tide_elevations(df.Longitude, df.Latitude,
    df.JulDay*86400.0, DIRECTORY=tide_dir, MODEL='FES2022',
    CROP=True, EPOCH=(1950,1,1,0,0,0), INFER_MINOR=True)
``` 

The RMS between `df.Pure_Tide` and `h` should be less than 2cm.

**Reply by fowlerovski** [reply]
Thank you @tsutterley. 
I appreciate your help very much.

EDIT: is there such a test for the loads as well? 
I see a lot of examples on pure tides, and similar, but no cases for the tide loads (surface displacement). 
Maybe it's so simple it doesn't need an example, or ground truth?

**Reply by tsutterley** [reply]
The `Geo_Tide` column in the test dataset should be `Pure_Tide + Load_Tide`, so you should be able to compare your load tide results against `Geo_Tide - Pure_Tide`

**Comment by fowlerovski** [comment]
Thanks for your help @tsutterley. 
I failed to succeed with pyTMD, but managed to directly access the netCDF4 format and ran from there.
Many thanks!

**Reply by tsutterley** [reply]
Anytime! Sorry you got stuck. Let me know if you need anything.

**Comment by dgstead** [comment]
New to using FES2022 extrapolated.
Does anyone know the timezone for exported lat/lon.
Not able to find whether it is native UTC - or provided as adjusted to the local date/time based on specific lat/lon coordinates. 
Searched sites and handbook - no references.

**Reply by tsutterley** [reply]
All of the calculations in pyTMD use times in UTC [(not technically but practically)](https://pytmd.readthedocs.io/en/latest/background/Time.html). FES2022 provides maps of phase lags w.r.t. the equilibrium tide phase, which I calculate in Terrestrial Time (TT).

**Reply by dgstead** [reply]
Thanks @tsutterley.
I'm using eo-tides for the moment, rather than pyTMD, but I know they operate in similar ways.

Not sure whether the output (of eo-tides or pyTMD) requires adjustment to the local date/time (UTC +11hrs in my case), or if the date/time is automatically adjusted by the tools on the basis of the precision lon/lat provided.

**Reply by robbibt** [reply]
In `eo-tides` I just pass times through directly to underlying pyTMD functions, so all times should be provided in UTC as well. 🙂

https://geoscienceaustralia.github.io/eo-tides/api/#eo_tides.model.model_tides
![Screenshot_20251106-164731.png](https://github.com/user-attachments/assets/af625ebe-d5bb-4146-a18f-bc9903d17c8e)

**Reply by dgstead** [reply]
Thanks Robbi.
